### ISSUE TITLE
Port twisted.python.failure to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5931]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5931|
|Type|enhancement|
|Created|2012-08-31 16:20:48Z|
|Branch|https://github.com/twisted/twisted/tree/failure-python3-5931-2|

Many parts of Twisted depend on `twisted.python.failure`.  Perhaps the debug functionality can be skipped for the initial pass, but most or all of the rest of the functionality will probably be required fairly soon.

This is a dependency for porting even a minimal subset of trial ([#5874](https://github.com/twisted/twisted/issues/5874)).

Some parts of the [#5802](https://github.com/twisted/twisted/issues/5802) branch may be useful in this effort.

<details><summary>Searchable metadata</summary>

```
trac-id__5931 5931
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_failure_python3_5931_2 branches/failure-python3-5931-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1346430048000000 1346430048000000
changetime__1347029214000000 1347029214000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

This depends on [#5913](https://github.com/twisted/twisted/issues/5913) and

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

... [#5932](https://github.com/twisted/twisted/issues/5932).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [35528]) Branching to 'failure-python3-5931'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35543]) Branching to 'failure-python3-5931-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set status to `new`|
|-|-|

Ready for review; running tests at http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/failure-python3-5931-2

Some notes:

  1. I punted on a couple of the really hideous failing tests for `_findFailure`, and opened separate ticket referenced in the code. Didn't seem important enough to block progress.
  1. Generator tests are effectively unchanged, just moved back into main test module; I only changed `g.next()` to `next(g)` and suppressed one test on Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  Surprisingly unterrible, really.

  1. We probably should already have started a guide for application developers, enumerating the changes they might need to make (like different exception handling around `Failure.trap` calls).  Can you start one for this ticket and put this item in it?  Then I guess we should go back and add stuff to it for all the tickets we resolved already.
  1. `twisted/test/test_failure.py`, _On both Python 2 and Python 3, the underlying exceptio is passed on:_ - missing an **n**
  1. `twisted/test/test_failure.py` needs a lot of help overall :/  Can you file tickets for:
    1. Improving the docstring for `test_notTrapped`
    1. Splitting up `test_printingSmokeTest` into separate tests for each method it is calling, as well as adding some more serious assertions
    1. The same for `test_printingCapturedVarsSmokeTest`
    1. Making `failure.startDebugMode` work in a less bad way.  We should do something that doesn't involve monkey-patching `Failure`.  (See next point)
  1. The change to `TestDebugMode.setUp` is wrong for Python 2.  `Failure.__dict__['__init__']` is the actual function object, `Failure.__init__` is an unbound method.  With the change, `failure.Failure.__init__ = origInit` now puts an unbound method into `Failure.__dict__`.
  1. Seems like the `inlineCallbacks` tests are in the wrong place.  They'd make more sense in a Deferred-related test module (in particular, not importing `twisted.internet` to test `twisted.python` would make a lot of sense).  Not sure this is really actionable here, but maybe there's another ticket to be filed for it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [35547]) Address review comments 2 and 4. Refs [#5931](https://github.com/twisted/twisted/issues/5931)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Opened [#5952](https://github.com/twisted/twisted/issues/5952), [#5953](https://github.com/twisted/twisted/issues/5953), [#5954](https://github.com/twisted/twisted/issues/5954) and [#5955](https://github.com/twisted/twisted/issues/5955) to address review comments 3 and 5.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [35548]) Howto covering python 3 porting, starting with failure. Refs [#5931](https://github.com/twisted/twisted/issues/5931)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, addressed review comments.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

And a new build started: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/failure-python3-5931-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @itamarst|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set status to `closed`|
|-|-|

(In [35549]) Merge failure-python3-5931-2

Author: itamar, vperic
Review: exarkun
Fixes: [#5931](https://github.com/twisted/twisted/issues/5931)

Port twisted.python.failure to Python 3, and add new porting howto.
